File: java/Diamond.java
interface Named {
  String getName();
}

interface Greeter extends Named {
  default void greet() {
    System.out.println("Hi. I'm " + getName());
  }
}
interface Polite extends Named {
  default void greet() {
    System.out.println("Let me introduce myself: My name is " + getName());
  }
}

class Person implements Greeter, Polite {
  String name;

  // won't compile if the method is not redefined
  public void greet() {
      Polite.super.greet();
      Greeter.super.greet();
  }

  Person(String name) {
    this.name = name;
  }

  public String getName() {
    return name;
  }
}

public class Diamond {
  public static void main(String[] args) {
    new Person("Georges").greet();
  }
}

File: ruby/diamond.rb
module Named
  def initialize(name)
    @name = name
  end

  def greet()
    puts "Hello my name is #{@name}"
  end
end

module Flying
  def initialize(wings)
    @wings = wings
  end

  def fly()
    puts "I'm flying with my #{@wings} wings"
  end
end

class Bird
  include Named
  include Flying

  def initialize(name, wings)
    super(wings)
    @name = name
  end
end

bird = Bird.new 'flappy', 2
bird.greet
bird.fly
